### DISCUSSION TITLE
Make webrtc-polyfill dependency optional

### DISCUSSION THREAD

**Discussion by proton-ab** [webtorrent/webtorrent#2795] [Ideas]
Depending on `webrtc-polyfill` pulls in `node-datachannel` which does not provide binaries for all platforms (namely, notice omission of support for `musl` and Alpine in general). Building `node-datachannel` requires at least `g++`, `cmake` and `openssl` as well as prebuild script which pulls over 600 additional unchecked dependencies via npm.

Since not all users require or benefit from WebRTC support, it would be beneficial to make this dependency optional, just like `utp-native` is currently optional.

**Comment by ThaUnknown** [comment]
my interpretation of https://github.com/webtorrent/webtorrent/issues/1734 is that the original goal of this change was to do exactly what webtorrent-hybrid does, but in the main package. Hybrid does not include WebRTC optionally, I actually don't use wrtc, but my own package built on top of ndc, which is miles faster and more compatible. So in this regard I believe I accomplished feross's original goal

**Reply by ThaUnknown** [reply]
Actually would simply making more prebuilds, or improving the build system in ndc fix this issue?

**Reply by funniray** [reply]
Hybrid doesn't include WebRTC optionally because hybrid itself was an option and only added WebRTC support. WebRTC on nodejs WebTorrent clients (in my opinion) should be optional, as WebRTC shouldn't be required for WebTorrent to work. 

I don't believe node-datachannel needs additional prebuilds as nodejs doesn't officially support musl. Users who wish to use docker containers with native modules can simply use a separate build stage with build-essentials install and copy the node_modules to the final container.

**Reply by ThaUnknown** [reply]
> Hybrid doesn't include WebRTC optionally because hybrid itself was an option and only added WebRTC support. WebRTC on nodejs WebTorrent clients (in my opinion) should be optional, as WebRTC shouldn't be required for WebTorrent to work.

yeah that's speculation at best, as I said, I went with what was visible and what I could directly infer

I do think ndc could improve their builds, maybe not ndc, but as proton mentioned, they put an insane amt of dependencies for their build, in comparison to say `utp-native`

**Comment by ThaUnknown** [comment]
the end conclusion of this is: `no`

since `webtorrent-hybrid` is now deprecated, all of the sub-dependencies which used it will point to webtorrent itself, installing webtorrent and not getting webrtc would be breaking, this can't be an optional dependency. 

Your point with the massive prebuild is something I already thought about myself long before WebRTC was implemented, we'll look if it can be simplified and make a PR to the downstream dependency.

afaik Node.js doesn't officially support musl, which is why it's omitted from the prebuilds

**Comment by proton-ab** [comment]
> afaik Node.js doesn't officially support musl, which is why it's omitted from the prebuilds

Not sure where you got this information from. NodeJS provides official Docker containers for Alpine and has multiple instances of documented behavioral differences (such as https://nodejs.org/api/fs.html#fspromisesrealpathpath-options) when running with musl.

**Reply by ThaUnknown** [reply]
yeah, they have code for it etc, but the support is not official, at least from my understanding, feel free to correct me, as musl is part of their unofficial builds, see: https://unofficial-builds.nodejs.org/download/release/v18.3.0/

**Reply by proton-ab** [reply]
You are correct that there are no prebuilts and that musl is considered Experimental as per https://github.com/nodejs/node/blob/main/BUILDING.md#platform-list, however this state of things has been questioned many times over the years considering how widely deployed Alpine docker images are and my personal opinion here is that this does not represent factual information and is more of byproduct of not enough volunteers guaranteeing support rather than anything else.

I have no qualms with `node-datachannel` if they do not wish to provide builds for musl due to its Experimental status, however I would also point out that they currently provide support for Windows x86 which is considered Experimental for compiling (ie. N-API) as per https://github.com/nodejs/node/blob/v16.0.0/BUILDING.md#fn6.

**Reply by ThaUnknown** [reply]
Fair fair, they actually supported down to node 12 before I forced them to do node 16, it's likely they don't even know that musl builds exist, you could likely ask them to make a prebuild for musl and they'll do it [later than sooner....]

**Comment by ThaUnknown** [comment]
@proton-ab node-datachannel now includes musl prebuilds, so this should be mostly resolved for you
